File: solutions/clojure/nucleotide-count/1/src/nucleotide_count.clj
(ns nucleotide-count)

(def valid-nucleotides #{\A \C \G \T})

(defn count-of-nucleotide-in-strand
  "Returns the count of a specific nucleotide in the strand"
  [nucleotide strand]
  (if (valid-nucleotides nucleotide)
    (count (filter #(= % nucleotide) strand))
    (throw (IllegalArgumentException. "Invalid nucleotide"))))


(defn nucleotide-counts [strand] ;; <- Arglist goes here
  ;; your code goes here
  (if (every? valid-nucleotides strand)
    (reduce (fn [counts n]
              (update counts n inc))
            {\A 0 \C 0 \G 0 \T 0}
            strand)
    (throw (IllegalArgumentException. "Invalid")))
)

File: solutions/javascript/nucleotide-count/1/nucleotide-count.js

export function countNucleotides(strand) {
  if (/.*[^ACGT].*/.test(strand)) {
    throw new Error("Invalid nucleotide in strand");
  }
  let nucleotideCount = {
    A: 0,
    C: 0,
    G: 0,
    T: 0,
  };
  for (let i = 0; i < strand.length; i++) {
    nucleotideCount[strand[i]]++;
  }
  let result = "";
  for (const key in nucleotideCount) {
    result += nucleotideCount[key] + " ";
  }
  return result.trim();
}